### ISSUE TITLE
build cphalcon 2.1.x failed using zephir 0.9.2a-dev

### ISSUE THREAD

**Issue by toddlxt** [zephir-lang/zephir#1244]
I was using Zephir 0.9.2a-dev to build cphalcon 2.1.x branch (CentOS 6.7, PHP 7.0.5, VirtualBox):

```
git clone http://github.com/phalcon/cphalcon
cd cphalcon
git checkout 2.1.x
zephir build --backend=ZendEngine3
```

And I got:

```
/usr/local/bin/zephir: line 26:  9793 
Segmentation fault (core dumped) php -d safe_mode=Off -d enable_dl=On $ZEPHIRDIR/compiler.php $*
```

Then I used `gdb php -c core.9793` to backtrace the error:

```
(gdb) bt
#0  zend_mm_free_heap (ptr=0x587) at /usr/src/debug/php-7.0.5/Zend/zend_alloc.c:1408
#1  _efree (ptr=0x587) at /usr/src/debug/php-7.0.5/Zend/zend_alloc.c:2466
#2  0x08d45f75 in parser_free_memory () at /path/to/zephir/parser/parser/parser.c:9345
#3  0x08d4d689 in xx_parse_program (
    program=0xb702a010 "\n/*\n +", '-' <repeats 72 times>, "+\n | Phalcon Framework", ' ' <repeats 54 times>, "|\n +", '-' <repeats 42 times>..., program_length=21835, 
    file_path=0xb75937e0 "/path/to/cphalcon/phalcon/acl/adapter/memory.zep")
    at /home/vagrant/zephir/parser/parser/parser.c:9877
#4  0x08d45e7b in zif_zephir_parse_file (execute_data=0xb7417b30, return_value=0xb74179d0)
    at /home/vagrant/zephir/parser/zephir_parser.c:48
#5  0x0828694a in execute_internal (execute_data=0xb7417b30, return_value=0xb74179d0)
    at /usr/src/debug/php-7.0.5/Zend/zend_execute.c:2036
...(Too long)
```

It seems the error happens in zephir/parser/parser/base.c line 121:

```
efree(parser_memory_manager->slots[i]);
```

This line was added two days ago, for "Fixed memory leaks".

I don't know what to do now. I just want to build phalcon.so with PHP 7.

**Comment by andresgutierrez** [comment]
Can you please try again using latest master?

**Comment by steffengy** [comment]
@andresgutierrez 

The travis PHP7 build still seems to segfault.

**Comment by sergeyklay** [comment]
Compile custom extension with latest Zephir and PHP 7:

```
mmap() failed: [12] Cannot allocate memory

Segmentation fault (core dumped)

Reading symbols from /usr/bin/php...(no debugging symbols found)...done.
[New LWP 2007]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/usr/bin/php7.0 -dextension=/usr/share/zephir/Library/../parser/zephir_parser.s'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007f6d289b326a in zend_mm_gc ()
(gdb) (gdb) #0  0x00007f6d289b326a in zend_mm_gc ()
No symbol table info available.
#1  0x00007f6d289b3bbc in ?? ()
No symbol table info available.
#2  0x00007f6d289b3e4d in ?? ()
No symbol table info available.
#3  0x00007f6d28a1396f in smart_str_erealloc ()
No symbol table info available.
#4  0x00007f6d17271482 in ?? () from /usr/lib/php/20151012/json.so
No symbol table info available.
#5  0x00007f6d17272b06 in php_json_encode_zval () from /usr/lib/php/20151012/json.so
No symbol table info available.
#6  0x00007f6d17271edf in ?? () from /usr/lib/php/20151012/json.so
No symbol table info available.
#7  0x00007f6d1727295c in php_json_encode_zval () from /usr/lib/php/20151012/json.so
No symbol table info available.
#8  0x00007f6d17271edf in ?? () from /usr/lib/php/20151012/json.so
No symbol table info available.
#9  0x00007f6d1727295c in php_json_encode_zval () from /usr/lib/php/20151012/json.so
No symbol table info available.
#10 0x00007f6d17271edf in ?? () from /usr/lib/php/20151012/json.so
No symbol table info available.
#11 0x00007f6d1727295c in php_json_encode_zval () from /usr/lib/php/20151012/json.so
No symbol table info available.
#12 0x00007f6d17271edf in ?? () from /usr/lib/php/20151012/json.so
No symbol table info available.
#13 0x00007f6d1727295c in php_json_encode_zval () from /usr/lib/php/20151012/json.so
No symbol table info available.
#14 0x00007f6d17271edf in ?? () from /usr/lib/php/20151012/json.so
No symbol table info available.
#15 0x00007f6d1727295c in php_json_encode_zval () from /usr/lib/php/20151012/json.so
No symbol table info available.
#16 0x00007f6d172706ab in ?? () from /usr/lib/php/20151012/json.so
No symbol table info available.
#17 0x00007f6d289c853b in dtrace_execute_internal ()
No symbol table info available.
#18 0x00007f6d2619ad61 in xdebug_execute_internal (current_execute_data=0x7f6d26616e00, return_value=0x7f6d26616c20) at /tmp/pear/temp/xdebug/xdebug.c:2035
        edata = <optimized out>
        fse = 0x7f6d2a55dfb0
        do_return = 0
        function_nr = 409
        restore_error_handler_situation = 0
        tmp_error_cb = 0x0
#19 0x00007f6d28a517e4 in ?? ()
No symbol table info available.
#20 0x00007f6d28a13bbb in execute_ex ()
No symbol table info available.
#21 0x00007f6d289c8439 in dtrace_execute_ex ()
No symbol table info available.
#22 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26616940) at /tmp/pear/temp/xdebug/xdebug.c:1890
        op_array = 0x7f6d26790528
        edata = <optimized out>
        dummy = <optimized out>
        fse = 0x7f6d2a55e240
        xfse = <optimized out>
        magic_cookie = <optimized out>
        do_return = 0
        function_nr = 400
        le = <optimized out>
#23 0x00007f6d28a514b0 in ?? ()
No symbol table info available.
#24 0x00007f6d28a13bbb in execute_ex ()
No symbol table info available.
#25 0x00007f6d289c8439 in dtrace_execute_ex ()
No symbol table info available.
#26 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26615f70) at /tmp/pear/temp/xdebug/xdebug.c:1890
        op_array = 0x7f6d26790ba8
        edata = <optimized out>
        dummy = <optimized out>
        fse = 0x7f6d2a55e390
        xfse = <optimized out>
        magic_cookie = <optimized out>
        do_return = 0
        function_nr = 399
        le = <optimized out>
#27 0x00007f6d28a514b0 in ?? ()
No symbol table info available.
#28 0x00007f6d28a13bbb in execute_ex ()
No symbol table info available.
#29 0x00007f6d289c8439 in dtrace_execute_ex ()
No symbol table info available.
#30 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26615bf0) at /tmp/pear/temp/xdebug/xdebug.c:1890
        op_array = 0x7f6d26612090
        edata = <optimized out>
        dummy = <optimized out>
        fse = 0x7f6d2a55dd30
        xfse = <optimized out>
        magic_cookie = <optimized out>
        do_return = 0
        function_nr = 375
        le = <optimized out>
#31 0x00007f6d28a514b0 in ?? ()
No symbol table info available.
#32 0x00007f6d28a13bbb in execute_ex ()
No symbol table info available.
#33 0x00007f6d289c8439 in dtrace_execute_ex ()
No symbol table info available.
#34 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d266159e0) at /tmp/pear/temp/xdebug/xdebug.c:1890
        op_array = 0x7f6d26612230
        edata = <optimized out>
        dummy = <optimized out>
        fse = 0x7f6d2a6c33d0
        xfse = <optimized out>
        magic_cookie = <optimized out>
        do_return = 0
        function_nr = 334
        le = <optimized out>
#35 0x00007f6d28a514b0 in ?? ()
No symbol table info available.
#36 0x00007f6d28a13bbb in execute_ex ()
No symbol table info available.
#37 0x00007f6d289c8439 in dtrace_execute_ex ()
No symbol table info available.
#38 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d266146a0) at /tmp/pear/temp/xdebug/xdebug.c:1890
        op_array = 0x7f6d2678b5c8
        edata = <optimized out>
        dummy = <optimized out>
        fse = 0x7f6d2a55d160
        xfse = <optimized out>
        magic_cookie = <optimized out>
        do_return = 0
        function_nr = 320
        le = <optimized out>
#39 0x00007f6d28a514b0 in ?? ()
No symbol table info available.
#40 0x00007f6d28a13bbb in execute_ex ()
No symbol table info available.
#41 0x00007f6d289c8439 in dtrace_execute_ex ()
No symbol table info available.
#42 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26614040) at /tmp/pear/temp/xdebug/xdebug.c:1890
        op_array = 0x7f6d2678b698
        edata = <optimized out>
        dummy = <optimized out>
        fse = 0x7f6d2a55d420
        xfse = <optimized out>
        magic_cookie = <optimized out>
        do_return = 0
        function_nr = 312
        le = <optimized out>
#43 0x00007f6d28a514b0 in ?? ()
No symbol table info available.
#44 0x00007f6d28a13bbb in execute_ex ()
No symbol table info available.
#45 0x00007f6d289c8439 in dtrace_execute_ex ()
No symbol table info available.
#46 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26613cb0) at /tmp/pear/temp/xdebug/xdebug.c:1890
        op_array = 0x7f6d2678b908
        edata = <optimized out>
        dummy = <optimized out>
        fse = 0x7f6d2a55d9d0
        xfse = <optimized out>
        magic_cookie = <optimized out>
        do_return = 0
        function_nr = 298
        le = <optimized out>
#47 0x00007f6d28a514b0 in ?? ()
No symbol table info available.
#48 0x00007f6d28a13bbb in execute_ex ()
No symbol table info available.
#49 0x00007f6d289c8439 in dtrace_execute_ex ()
No symbol table info available.
#50 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26613c30) at /tmp/pear/temp/xdebug/xdebug.c:1890
        op_array = 0x7f6d2678bc48
        edata = <optimized out>
        dummy = <optimized out>
        fse = 0x7f6d2a55cf50
        xfse = <optimized out>
        magic_cookie = <optimized out>
        do_return = 0
        function_nr = 297
        le = <optimized out>
#51 0x00007f6d28a514b0 in ?? ()
No symbol table info available.
#52 0x00007f6d28a13bbb in execute_ex ()
No symbol table info available.
#53 0x00007f6d289c8439 in dtrace_execute_ex ()
No symbol table info available.
#54 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26613940) at /tmp/pear/temp/xdebug/xdebug.c:1890
        op_array = 0x7f6d26604ef8
        edata = <optimized out>
        dummy = <optimized out>
        fse = 0x7f6d2a6cc420
        xfse = <optimized out>
        magic_cookie = <optimized out>
        do_return = 0
        function_nr = 224
        le = <optimized out>
#55 0x00007f6d28a514b0 in ?? ()
No symbol table info available.
#56 0x00007f6d28a13bbb in execute_ex ()
No symbol table info available.
#57 0x00007f6d289c8439 in dtrace_execute_ex ()
No symbol table info available.
#58 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d266130b0) at /tmp/pear/temp/xdebug/xdebug.c:1890
        op_array = 0x7f6d26603798
        edata = <optimized out>
        dummy = <optimized out>
        fse = 0x7f6d2a6c3140
        xfse = <optimized out>
        magic_cookie = <optimized out>
        do_return = 0
        function_nr = 15
        le = <optimized out>
#59 0x00007f6d28a514b0 in ?? ()
No symbol table info available.
#60 0x00007f6d28a13bbb in execute_ex ()
No symbol table info available.
#61 0x00007f6d289c8439 in dtrace_execute_ex ()
No symbol table info available.
#62 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26613030) at /tmp/pear/temp/xdebug/xdebug.c:1890
        op_array = 0x7f6d2667a000
        edata = <optimized out>
        dummy = <optimized out>
        fse = 0x7f6d2a6a29a0
        xfse = <optimized out>
        magic_cookie = <optimized out>
        do_return = 0
        function_nr = 0
        le = <optimized out>
#63 0x00007f6d28a5cc5f in zend_execute ()
No symbol table info available.
#64 0x00007f6d289d7a84 in zend_execute_scripts ()
No symbol table info available.
#65 0x00007f6d2897b8e8 in php_execute_script ()
No symbol table info available.
#66 0x00007f6d28a5e839 in ?? ()
No symbol table info available.
#67 0x00007f6d2886b620 in main ()
No symbol table info available.
(gdb) rax            0x80020000 2147614720
rbx            0x7f6d26600040   140106771988544
rcx            0x20000  131072
rdx            0x7f6d00001c07   140106128169991
rsi            0x7f6d00000000   140106128162816
rdi            0x6  6
rbp            0x7f6d26600060   0x7f6d26600060
rsp            0x7ffdb38a6000   0x7ffdb38a6000
r8             0x0  0
r9             0x200    512
r10            0x1  1
r11            0x7f6d26600060   140106771988576
r12            0x0  0
r13            0x7f6d0dc15000   140106358935552
r14            0x400000 4194304
r15            0x2f4778 3098488
rip            0x7f6d289b326a   0x7f6d289b326a <zend_mm_gc+138>
eflags         0x10206  [ PF IF RF ]
cs             0x33 51
ss             0x2b 43
ds             0x0  0
es             0x0  0
fs             0x0  0
gs             0x0  0
(gdb) => 0x7f6d289b326a <zend_mm_gc+138>:   cmp    (%rsi),%rbx
   0x7f6d289b326d <zend_mm_gc+141>: jne    0x7f6d289b3601 <zend_mm_gc+1057>
   0x7f6d289b3273 <zend_mm_gc+147>: mov    %rdx,%rax
   0x7f6d289b3276 <zend_mm_gc+150>: and    $0x1fffff,%eax
   0x7f6d289b327b <zend_mm_gc+155>: shr    $0xc,%rax
   0x7f6d289b327f <zend_mm_gc+159>: mov    0x1f8(%rsi,%rax,4),%ecx
   0x7f6d289b3286 <zend_mm_gc+166>: movslq %eax,%rdi
   0x7f6d289b3289 <zend_mm_gc+169>: test   $0x40000000,%ecx
   0x7f6d289b328f <zend_mm_gc+175>: je     0x7f6d289b32a8 <zend_mm_gc+200>
   0x7f6d289b3291 <zend_mm_gc+177>: and    $0x1ff0000,%ecx
   0x7f6d289b3297 <zend_mm_gc+183>: mov    %eax,%edi
   0x7f6d289b3299 <zend_mm_gc+185>: shr    $0x10,%ecx
   0x7f6d289b329c <zend_mm_gc+188>: sub    %ecx,%edi
   0x7f6d289b329e <zend_mm_gc+190>: movslq %edi,%rdi
   0x7f6d289b32a1 <zend_mm_gc+193>: mov    0x1f8(%rsi,%rdi,4),%ecx
   0x7f6d289b32a8 <zend_mm_gc+200>: mov    %ecx,%eax
(gdb) 
Thread 1 (Thread 0x7f6d28cff740 (LWP 2007)):
#0  0x00007f6d289b326a in zend_mm_gc ()
#1  0x00007f6d289b3bbc in ?? ()
#2  0x00007f6d289b3e4d in ?? ()
#3  0x00007f6d28a1396f in smart_str_erealloc ()
#4  0x00007f6d17271482 in ?? () from /usr/lib/php/20151012/json.so
#5  0x00007f6d17272b06 in php_json_encode_zval () from /usr/lib/php/20151012/json.so
#6  0x00007f6d17271edf in ?? () from /usr/lib/php/20151012/json.so
#7  0x00007f6d1727295c in php_json_encode_zval () from /usr/lib/php/20151012/json.so
#8  0x00007f6d17271edf in ?? () from /usr/lib/php/20151012/json.so
#9  0x00007f6d1727295c in php_json_encode_zval () from /usr/lib/php/20151012/json.so
#10 0x00007f6d17271edf in ?? () from /usr/lib/php/20151012/json.so
#11 0x00007f6d1727295c in php_json_encode_zval () from /usr/lib/php/20151012/json.so
#12 0x00007f6d17271edf in ?? () from /usr/lib/php/20151012/json.so
#13 0x00007f6d1727295c in php_json_encode_zval () from /usr/lib/php/20151012/json.so
#14 0x00007f6d17271edf in ?? () from /usr/lib/php/20151012/json.so
#15 0x00007f6d1727295c in php_json_encode_zval () from /usr/lib/php/20151012/json.so
#16 0x00007f6d172706ab in ?? () from /usr/lib/php/20151012/json.so
#17 0x00007f6d289c853b in dtrace_execute_internal ()
#18 0x00007f6d2619ad61 in xdebug_execute_internal (current_execute_data=0x7f6d26616e00, return_value=0x7f6d26616c20) at /tmp/pear/temp/xdebug/xdebug.c:2035
#19 0x00007f6d28a517e4 in ?? ()
#20 0x00007f6d28a13bbb in execute_ex ()
#21 0x00007f6d289c8439 in dtrace_execute_ex ()
#22 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26616940) at /tmp/pear/temp/xdebug/xdebug.c:1890
#23 0x00007f6d28a514b0 in ?? ()
#24 0x00007f6d28a13bbb in execute_ex ()
#25 0x00007f6d289c8439 in dtrace_execute_ex ()
#26 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26615f70) at /tmp/pear/temp/xdebug/xdebug.c:1890
#27 0x00007f6d28a514b0 in ?? ()
#28 0x00007f6d28a13bbb in execute_ex ()
#29 0x00007f6d289c8439 in dtrace_execute_ex ()
#30 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26615bf0) at /tmp/pear/temp/xdebug/xdebug.c:1890
#31 0x00007f6d28a514b0 in ?? ()
#32 0x00007f6d28a13bbb in execute_ex ()
#33 0x00007f6d289c8439 in dtrace_execute_ex ()
#34 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d266159e0) at /tmp/pear/temp/xdebug/xdebug.c:1890
#35 0x00007f6d28a514b0 in ?? ()
#36 0x00007f6d28a13bbb in execute_ex ()
#37 0x00007f6d289c8439 in dtrace_execute_ex ()
#38 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d266146a0) at /tmp/pear/temp/xdebug/xdebug.c:1890
#39 0x00007f6d28a514b0 in ?? ()
#40 0x00007f6d28a13bbb in execute_ex ()
#41 0x00007f6d289c8439 in dtrace_execute_ex ()
#42 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26614040) at /tmp/pear/temp/xdebug/xdebug.c:1890
#43 0x00007f6d28a514b0 in ?? ()
#44 0x00007f6d28a13bbb in execute_ex ()
#45 0x00007f6d289c8439 in dtrace_execute_ex ()
#46 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26613cb0) at /tmp/pear/temp/xdebug/xdebug.c:1890
#47 0x00007f6d28a514b0 in ?? ()
#48 0x00007f6d28a13bbb in execute_ex ()
#49 0x00007f6d289c8439 in dtrace_execute_ex ()
#50 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26613c30) at /tmp/pear/temp/xdebug/xdebug.c:1890
#51 0x00007f6d28a514b0 in ?? ()
#52 0x00007f6d28a13bbb in execute_ex ()
#53 0x00007f6d289c8439 in dtrace_execute_ex ()
#54 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26613940) at /tmp/pear/temp/xdebug/xdebug.c:1890
#55 0x00007f6d28a514b0 in ?? ()
#56 0x00007f6d28a13bbb in execute_ex ()
#57 0x00007f6d289c8439 in dtrace_execute_ex ()
#58 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d266130b0) at /tmp/pear/temp/xdebug/xdebug.c:1890
#59 0x00007f6d28a514b0 in ?? ()
#60 0x00007f6d28a13bbb in execute_ex ()
#61 0x00007f6d289c8439 in dtrace_execute_ex ()
#62 0x00007f6d2619a3ef in xdebug_execute_ex (execute_data=0x7f6d26613030) at /tmp/pear/temp/xdebug/xdebug.c:1890
#63 0x00007f6d28a5cc5f in zend_execute ()
#64 0x00007f6d289d7a84 in zend_execute_scripts ()
#65 0x00007f6d2897b8e8 in php_execute_script ()
#66 0x00007f6d28a5e839 in ?? ()
#67 0x00007f6d2886b620 in main ()
```

**Comment by toddlxt** [comment]
@andresgutierrez Successfully compiled cpahlcon 2.1.x using the latest master. Thank you very much!

**Comment by mruz** [comment]
After [PhalconPHP and PHP7](http://phalcon.io/phalconphp-and-php7) article a lots of people trying to build phalcon 2.1 on php7, but there are not fixed [php7 issues](https://github.com/phalcon/zephir/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+PHP7+), so it could compile but is very unstable.

> After a few months we are finally delivered PhalconPHP 2.1 and Zephir 0.9.x which brings PHP7 support.

This is not true, Phalcon 2.1 and fully php7 support are not delivered yet, so article misleading.

**Comment by toddlxt** [comment]
@mruz That's true. When I saw "Phalcon 2.1.x has beta PHP7 support", I couldn't wait to download and compile 2.1.x. However, php crashed and nginx returned 502 page when I used Models or PHQL to retrieve data from Mysql (\mysqli worked fine, and the same code worked fine with 2.0.10). I still feel very excited about the coming PHP7 support, though. I admire the authors of Phalcon and Zephir, and believe they will resolve these bugs soon. I know phalcon 2.1.x is a beta version, and I'm willing to test it and submit issues to help improve Phalcon.
